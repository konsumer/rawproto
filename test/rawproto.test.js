import { getData, getProto } from '../index.js'
import protobuf from 'protobufjs'
import hexy from 'hexy'

// get proto-def for tests
const proto = await protobuf.load(new URL('demo.proto', import.meta.url).pathname)

const Test = proto.lookupType('Test')
const Test2 = proto.lookupType('Test2')

// I just define this once to save some code
const demoObject = {
  nums: [1, 2, 3, 4, 5],
  num: 1,
  str: 'hello',
  children: [
    { num: 1, str: 'cool', children: [{ num: 1 }] },
    { num: 2, str: 'awesome', children: [{ num: 2 }] },
    { num: 3, str: 'neat', children: [{ num: 3 }] }
  ],
  extra: 'this one is extra'
}

// make some encoded messages for testing
const test1 = Test.encode(demoObject).finish()
const test2 = Test2.encode(demoObject).finish()

// console.log(hexy.hexy(test1))
// console.log(hexy.hexy(test2))

describe('rawproto', () => {
  test('Test raw', () => {
    expect(getData(test1)).toMatchSnapshot()
  })

  test('Test2 raw', () => {
    expect(getData(test2)).toMatchSnapshot()
  })

  test('Some missing fields', () => {
    expect(getData(test2, Test)).toMatchSnapshot()
  })

  test('getProto', () => {
    const p = getProto(test2, Test2)
    // console.log(p)
    expect(p).toMatchSnapshot()
  })
})
