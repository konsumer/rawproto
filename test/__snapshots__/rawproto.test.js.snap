// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rawproto > Some missing fields 1`] = `
[
  {
    "1": {
      "data": [
        1,
        2,
        3,
        4,
        5,
      ],
      "type": "Buffer",
    },
  },
  {
    "2": 1,
  },
  {
    "3": "hello",
  },
  {
    "4": [
      {
        "1": 1,
      },
      {
        "2": "cool",
      },
      {
        "3": [
          {
            "1": 1,
          },
        ],
      },
    ],
  },
  {
    "4": [
      {
        "1": 2,
      },
      {
        "2": "awesome",
      },
      {
        "3": [
          {
            "1": 2,
          },
        ],
      },
    ],
  },
  {
    "4": [
      {
        "1": 3,
      },
      {
        "2": "neat",
      },
      {
        "3": [
          {
            "1": 3,
          },
        ],
      },
    ],
  },
  {
    "5": "this one is extra",
  },
]
`;

exports[`rawproto > Test raw 1`] = `
[
  {
    "1": {
      "data": [
        1,
        2,
        3,
        4,
        5,
      ],
      "type": "Buffer",
    },
  },
  {
    "2": 1,
  },
  {
    "3": "hello",
  },
  {
    "4": [
      {
        "1": 1,
      },
      {
        "2": "cool",
      },
      {
        "3": [
          {
            "1": 1,
          },
        ],
      },
    ],
  },
  {
    "4": [
      {
        "1": 2,
      },
      {
        "2": "awesome",
      },
      {
        "3": [
          {
            "1": 2,
          },
        ],
      },
    ],
  },
  {
    "4": [
      {
        "1": 3,
      },
      {
        "2": "neat",
      },
      {
        "3": [
          {
            "1": 3,
          },
        ],
      },
    ],
  },
]
`;

exports[`rawproto > Test2 raw 1`] = `
[
  {
    "1": {
      "data": [
        1,
        2,
        3,
        4,
        5,
      ],
      "type": "Buffer",
    },
  },
  {
    "2": 1,
  },
  {
    "3": "hello",
  },
  {
    "4": [
      {
        "1": 1,
      },
      {
        "2": "cool",
      },
      {
        "3": [
          {
            "1": 1,
          },
        ],
      },
    ],
  },
  {
    "4": [
      {
        "1": 2,
      },
      {
        "2": "awesome",
      },
      {
        "3": [
          {
            "1": 2,
          },
        ],
      },
    ],
  },
  {
    "4": [
      {
        "1": 3,
      },
      {
        "2": "neat",
      },
      {
        "3": [
          {
            "1": 3,
          },
        ],
      },
    ],
  },
  {
    "5": "this one is extra",
  },
]
`;

exports[`rawproto > getProto 1`] = `
"syntax = \\"proto3\\";

message MessageRoot {
  bytes field1 = 1; // could be a repeated-value, string, bytes, or malformed sub-message
  int32 field2 = 2; // could be a int32, int64, uint32, bool, enum, etc, or even a float of some kind
  string field3 = 3; // could be a repeated-value, string, bytes, or malformed sub-message

  message Message4 {
    int32 field1 = 1; // could be a int32, int64, uint32, bool, enum, etc, or even a float of some kind
    string field2 = 2; // could be a repeated-value, string, bytes, or malformed sub-message

    message Message3 {
      int32 field1 = 1; // could be a int32, int64, uint32, bool, enum, etc, or even a float of some kind
    }
    
    Message3 subMessage3 = 3;
  }
  
  repeated Message4 subMessage4 = 4;
  string field5 = 5; // could be a repeated-value, string, bytes, or malformed sub-message
}"
`;
